src/metaopt/io/database/mongodb.py

```diff
@@ -0,0 +1,137 @@
+import json
+import bson
+import pymongo
+
+from pymongo import MongoClient
+from datetime import datetime
+from abstractdb import AbstractDB
+
+try:
+    from StringIO import StringIO
+except ImportError:
+    from io import StringIO
+
+try:
+    # Python 3.x
+    from urllib.parse import quote_plus
+except ImportError:
+    # Python 2.x
+    from urllib import quote_plus
+
+
+class MongoDB(AbstractDB):
+    def __init__(self, dbhost='localhost', dbname='metaopt', dbport=27017, \
+                                            username=None, password=None):
+        self.dbhost = str(dbhost)
+        self.dbname = str(dbname)
+        self.username = str(username)
+        self.password = str(password)
+
+        # we might have a different port for the database port
+        # so here we do a check if the port is there
+
+        self.dbport = str(dbport)
+
+        # set the connection string
+        # this should be mongodb://mongodb0.example.net:27017
+        # according to MongoDB documentation
+        # https://docs.mongodb.com/getting-started/python/client/
+
+        self.uri = 'mongodb://' + self.dbhost + ':' + self.dbport
+        #self.uri = "mongodb://%s:%s@%s:%s" % (quote_plus(self.username), \
+        #                                      quote_plus(self.password), \
+        #                                     self.dbhost, self.dbport)
+        # try to initialize connection
+        # setting this as a function allows to make sure reconnection
+        # is properly done in the case of an unreliable network
+
+        self.conn = self.initiate_connection()
+        self.db = self.conn[self.dbname]
+
+    def initiate_connection(self):
+        try:
+            conn = MongoClient(self.uri)
+            self.conn = conn
+        except pymongo.errors.ConnectionFailure:
+            raise Exception("Could not connect to server")
```

Comment:
Could ConnectionFailure connect important information that we would like to pass on?

Reply:
Like the enumeration of what error type it would be? So I guess `raise Exeception("Exception caught %s, e)` instead?

Reply:
I don't see how we would need it? If `pymongo.errors.ConnectionFailure as e` we pass on (like in C), the assert value for the error is null, thus outside of the `try exception` block we would always return a valid connection object.

Reply:
By passing on, I meant passing the error information in the raised error so that it finds its way in the stack trace printed to the user.

Reply:
Hmmmn conn itself contains enough information to let the user know if things aren't right

Reply:
It is not a good idea to raise or catch pure Exception objects; there is no way to differentiate between cases. I agree with Dendi's thought to reraise an exception here, but it should be an Exception of our own w.r.t. common database problems, so that we can wrap in the future other libraries as well.

However, I think we should still keep the traceback though, Pythonic stuff. `raise wrapped_e from e` clauses are the case here, but unfortunately however beautiful (I find) they are, they are Py3 only. But there is this:
http://python-future.org/compatible_idioms.html#raising-exceptions (`from future.utils`)
https://pythonhosted.org/six/#six.raise_from (`from six` although they do not implement the correct impl of this functn; there's a current PR about this)

For truthful reference: https://www.python.org/dev/peps/pep-3134/#rationale
`.. seemore:: difference between implicit and explicit exception chaining :P`

Reply:
@tsirif I referred to this [example](https://github.com/mila-udem/blocks/blob/master/blocks/utils/utils.py#L32) in another comment for a way to reraise an exception. I agree we should not reraise it as a pure Exception. But again, I'm not sure we would need to reraise it. Isn't the ConnectionFailure explicit enough?

Reply:
Either both, I added it to the `__init__.py` in `utils.py` and redid all of the `exception` blocks.